### ISSUE TITLE
Transpile fails when multiple circuits with UnitaryGate or DiagonalGate

### ISSUE THREAD

**Issue by doichanj** [Qiskit/qiskit#12953]
### Environment

- **Qiskit version1.3: 
- **Python version**: 
- **Operating system**: 

### What is happening?

Transpile fails when the list of input circuits contains `UnitaryGate` or `DiagonalGate` with error as following:

```
pyo3_runtime.PanicException: internal error: entered unreachable code: 'num_qubits' property is reserved for normal operations only.
```
And single circuit input with `UnitaryGate` or `DiagonalGate` 

see https://github.com/Qiskit/qiskit-aer/actions/runs/10381689813/job/28743530145

### How can we reproduce the issue?

```
from qiskit_aer import AerSimulator
from qiskit import QuantumCircuit, transpile
from qiskit.quantum_info.random import random_unitary

backend = AerSimulator(method="statevector", device="CPU")

qc1 = QuantumCircuit(2)
qc1.unitary(random_unitary(4, seed=123), [0, 1])
# using followings instead of unitary is OK
#qc1.h(0)
#qc1.cx(0, 1)
qc1.measure_all()

qc2 = QuantumCircuit(2)
qc2.unitary(random_unitary(4, seed=123), [0, 1])
# using followings instead of unitary is OK
#qc2.h(1)
#qc2.cx(1, 0)
qc2.measure_all()

circuits = [qc1, qc2]
# circuits = [qc1]   this is also OK
circuits = transpile(circuits, backend)

result = backend.run(circuits).result()
print(result)
```

### What should happen?

No error in transpile

**Comment by jakelishman** [comment]
@raynelfss: this message is coming from the `Target` port - could you have a look into it?

Doi-san: I think you might be testing against Qiskit's `main` branch, rather than 1.2.0rc1?  The error message in question isn't part of the 1.2 release - it'll be in 1.3, though (it's #12292, and was the point we branched 1.2 away).

**Comment by doichanj** [comment]
yes, this error only occurs on main branch

**Comment by raynelfss** [comment]
This error occurs when attempting to get the `num_qubits` property from a variadic operation. I can change this behavior to return `None` instead of panicking. As Python would usually just return a `property` instance.

**Comment by raynelfss** [comment]
Hello @doichanj, what operating system are you on at the moment? The code snippet you provided doesn't seem to replicate the issue for me even when using the latest `main` in `qiskit-aer`.

**Comment by doichanj** [comment]
I'm using Ubuntu 22.04.4 for script above.
And for the test on github (https://github.com/Qiskit/qiskit-aer/actions/runs/10381689813/job/28743530145) also on Ubuntu 22.04.4 and uses combinations of the latest main branches of Qiskit and Qiskit Aer

**Comment by raynelfss** [comment]
Hello, I was finally able to replicate the issue on my end. But I had to go all the way back to the first commit that introduced Target in rust (https://github.com/Qiskit/qiskit/commit/898fdec48ab01ebee614ab7d32ebab597872ca20), and had to also turn on `QISKIT_PARALLEL`. The error does not happen on newer versions of Qiskit or `qiskit-aer` (because of https://github.com/Qiskit/qiskit-aer/commit/44bd7eba599bfa094b2ac1f0b74414338db9bdf4). 

The issue is fixed by #12957 and I'm currently adding a test-case based on what you sent, but since I can't determine what specific circumstances trigger this (due to serialization) it will take more time and thorough testing.
